File: src/MixNetFx/FacInterfaces/Legatro/frmLegatroTimeDataConfigDialog.vb
Imports System.Windows.Forms
Imports ActiveDev.Data.SqlClient
Imports Facesso.Data

Public Class frmLegatroTimeDataConfigDialog

    Private myTaskItem As LegatroTimeDataImport
    Private mySortedWorkgroupInfoItems As List(Of WorkGroupInfo)
    Private myWorksiteOrProjects As List(Of WorksitesOrProjects)

    Public Function HandleDialog(ByVal TaskItem As IFacessoImportTaskItem) As DialogResult
        myTaskItem = DirectCast(TaskItem, LegatroTimeDataImport)
        InitializeControls()
        If Not String.IsNullOrEmpty(myTaskItem.LegatroSQLConnectionString) Then
            txtSqlConnectionString.Text = myTaskItem.LegatroSQLConnectionString
            RebuildLists()
        End If

        Me.ShowDialog()
        Return Me.DialogResult
    End Function

    Private Sub btnSelectSqlConnection_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSelectSqlConnection.Click
        Dim frm As New ADDatabaseConnectionDialog
        Dim connBuilder = frm.GetConnectionBuilder("Legatro-Datenbankinstanz auswählen:")
        If connBuilder IsNot Nothing Then
            txtSqlConnectionString.Text = connBuilder.ConnectionString
            If myTaskItem IsNot Nothing Then
                myTaskItem.LegatroSQLConnectionString = txtSqlConnectionString.Text
            End If
            InitializeControls()
            RebuildLists()
        End If
    End Sub

    Private Sub InitializeControls()
        With lvwLegatroWorksitesOrProjects
            .Columns.Clear()
            With .Columns
                .Add("Nummer:")
                .Add("Arbeitsplatzname:")
            End With
        End With
    End Sub

    Private Sub RebuildLists()
        lvwLegatroWorksitesOrProjects.Items.Clear()
        tvwAssignments.Nodes.Clear()

        If Not String.IsNullOrEmpty(myTaskItem.LegatroSQLConnectionString) Then
            Dim dc As New LegatroDataContext(myTaskItem.LegatroSQLConnectionString)
            Try
                myWorksiteOrProjects = (From wsItem In dc.WorksitesOrProjects _
                           Where Not wsItem.IsProject _
                           Order By wsItem.WorkEntityNumber).ToList
            Catch ex As Exception
                MessageBox.Show("Beim Abrufen der Daten aus der Legatro-Datenbank ist ein Fehler aufgetreten." & vbNewLine & _
                                "Bitte Überprüfen Sie die Netzwerkverbindung und die Verbindungszeichenfolge zur Datenbank.", _
                                "Verbindungsaufbau zu Legatro nicht möglich!", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return
            End Try

            For Each wsItem In myWorksiteOrProjects
                Dim lvwItem As New ListViewItem(wsItem.WorkEntityNumber.ToString("000000"))
                lvwItem.SubItems.Add(wsItem.WorkEntityName)
                lvwItem.Tag = wsItem

                'Rausfinden, ob das Item schon zugewiesen wurde
                If myTaskItem.ConversionItems IsNot Nothing Then
                    For Each taskItem In myTaskItem.ConversionItems
                        If taskItem.AlienElementID = wsItem.WorkEntityNumber Then
                            '...dann in Fettschrift anzeigen
                            If taskItem.HomeElementID > -1 Then
                                lvwItem.Font = New Drawing.Font(lvwItem.Font, Drawing.FontStyle.Bold)
                            End If
                        End If
                    Next
                End If
                lvwLegatroWorksitesOrProjects.Items.Add(lvwItem)
            Next
        End If
        For Each colItem As ColumnHeader In lvwLegatroWorksitesOrProjects.Columns
            colItem.Width = -2
        Next

        Dim wgItems As New WorkGroupInfoItems(False)
        mySortedWorkgroupInfoItems = (From wgSortedItem In wgItems _
                      Order By wgSortedItem.WorkGroupNumber).ToList

        For Each item In mySortedWorkgroupInfoItems
            Dim tn = tvwAssignments.Nodes.Add(item.WorkGroupNumber & ": " & item.WorkGroupName)
            tn.Tag = item
            If item.IsActive Then
                tn.ForeColor = Drawing.Color.Blue
                tn.NodeFont = New Drawing.Font(Me.Font, Drawing.FontStyle.Bold)
            Else
                tn.ForeColor = Drawing.Color.Red
            End If
        Next
        Application.DoEvents()

        If myTaskItem.ConversionItems Is Nothing Then
            myTaskItem.ConversionItems = myTaskItem.AssembleConversionItems
        Else
            'Abgleich, ob sich was geändert hat
            Dim tmpItems = myTaskItem.AssembleConversionItems
            Dim itemsToAdd = New List(Of FacessoConversionItemBase)
            For Each tmpItem In tmpItems
                Dim found = False
                For Each taskItem In myTaskItem.ConversionItems
                    If taskItem.AlienElementID = tmpItem.AlienElementID Then
                        found = True
                        Exit For
                    End If
                Next
                If Not found Then
                    itemsToAdd.Add(tmpItem)
                End If
            Next
            For Each itemToAdd In itemsToAdd
                myTaskItem.ConversionItems.Add(itemToAdd)
            Next
        End If

        For Each item In myTaskItem.ConversionItems
            'Ein Facesso-Element (Worksite) ist zugeordnet bei >-1

            If item.HomeElementID > -1 Then
                'Home-Element im Tree suchen
                For Each nodeItem As TreeNode In tvwAssignments.Nodes
                    If DirectCast(nodeItem.Tag, WorkGroupInfo).WorkGroupNumber = item.HomeElementID Then
                        'Gefunden: Alien-Element und Item-Name als Zwei eintragen
                        Dim tmpNode = nodeItem.Nodes.Add(item.AlienElementID.ToString("000000") & ": " & item.Itemname)
                        tmpNode.Tag = item.HomeElementID
                        nodeItem.Expand()
                    End If
                Next
            End If
        Next
    End Sub

    Private Sub btnAdd_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnAdd.Click
        If lvwLegatroWorksitesOrProjects.SelectedItems.Count = 0 Then
            MessageBox.Show("Bitte wählen Sie einen Legatro-Arbeitsplatz zum Zuweisen aus!", _
                             "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End If

        If tvwAssignments.SelectedNode Is Nothing Then
            MessageBox.Show("Bitte wählen Sie eine Facesso-Produktiv-Site aus, der der Legatro-Arbeitsplatz zugewiesen werden soll!", _
                             "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End If

        Dim legItem = DirectCast(lvwLegatroWorksitesOrProjects.SelectedItems(0).Tag, WorksitesOrProjects)
        Dim facItem = DirectCast(tvwAssignments.SelectedNode.Tag, WorkGroupInfo)

        For Each cItem In myTaskItem.ConversionItems
            If cItem.AlienElementID = legItem.WorkEntityNumber Then
                cItem.HomeElementID = facItem.WorkGroupNumber
                cItem.HomeElementName = facItem.WorkGroupName
                Exit For
            End If
        Next

        RebuildLists()
    End Sub

    Private Sub btnRemove_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRemove.Click
        If tvwAssignments.SelectedNode Is Nothing Then
            MessageBox.Show("Bitte wählen Sie eine Facesso-Produktiv-Site aus, deren Zuweisung aufgehoben werden soll!", _
                             "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End If

        If tvwAssignments.SelectedNode.Parent Is Nothing And tvwAssignments.SelectedNode.Nodes.Count = 0 Then
            MessageBox.Show("Bitte wählen Sie eine Zweig aus, der ein Legatro-Arbeitsplatz/Kostenstelle zugeordnet ist!", _
                             "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End If

        Dim currentNode As TreeNode

        'Dafür sorgen, dass immer der Hauptzweig ausgewählt ist.
        If tvwAssignments.SelectedNode.Parent IsNot Nothing Then
            currentNode = tvwAssignments.SelectedNode.Parent
        Else
            currentNode = tvwAssignments.SelectedNode
        End If

        'Zugeordnetes WorkgroupInfo-Objekt finden
        Dim workgroupNumber = CInt(currentNode.Nodes(0).Tag)
        Dim facItem = (From fItem In mySortedWorkgroupInfoItems _
                      Where fItem.WorkGroupNumber = workgroupNumber).SingleOrDefault
        If facItem Is Nothing Then
            MessageBox.Show("Interner Zuordnungsfehler - bitte rufen Sie nach Verlassen diesen Dialog erneut auf.", _
                            "Interner Zuordnungsfehler", MessageBoxButtons.OK, MessageBoxIcon.Asterisk)
            Return
        End If

        'Sicherheitsabfrage:
        Dim dr = MessageBox.Show("Sind Sie sicher, dass Sie die Zuordnung zur Produktiv-Site " & facItem.DisplayName & " aufheben wollen?", _
                               "Zuordnung aufheben?", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button2)
        If dr = Windows.Forms.DialogResult.Yes Then

            For Each cItem In myTaskItem.ConversionItems
                If cItem.HomeElementID = facItem.WorkGroupNumber Then
                    cItem.HomeElementID = -1
                    cItem.HomeElementName = ""
                    Exit For
                End If
            Next
            RebuildLists()
        End If
    End Sub
End Class

File: src/NetFxCSharp/Facesso.Interfaces/Legatro/frmLegatroTimeDataConfigDialog.cs
using ActiveDev.Data.SqlClient;
using Facesso.Data;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.Interfaces
{
    public partial class frmLegatroTimeDataConfigDialog
    {
        private LegatroTimeDataImport myTaskItem;
        private List<WorkGroupInfo> mySortedWorkgroupInfoItems;
        private List<WorksitesOrProjects> myWorksiteOrProjects;
        public DialogResult HandleDialog(IFacessoImportTaskItem TaskItem)
        {
            myTaskItem = ((LegatroTimeDataImport)TaskItem);
            InitializeControls();
            if (!(string.IsNullOrEmpty(myTaskItem.LegatroSQLConnectionString)))
            {
                txtSqlConnectionString.Text = myTaskItem.LegatroSQLConnectionString;
                RebuildLists();
            }

            this.ShowDialog();
            return this.DialogResult;
        }

        private void btnSelectSqlConnection_Click(System.Object sender, System.EventArgs e)
        {
            ADDatabaseConnectionDialog frm = new ADDatabaseConnectionDialog();
            var connBuilder = frm.GetConnectionBuilder("Legatro-Datenbankinstanz auswählen:");
            if (connBuilder != null)
            {
                txtSqlConnectionString.Text = connBuilder.ConnectionString;
                if (myTaskItem != null)
                {
                    myTaskItem.LegatroSQLConnectionString = txtSqlConnectionString.Text;
                }

                InitializeControls();
                RebuildLists();
            }
        }

        private void InitializeControls()
        {
            {
                var __with0 = lvwLegatroWorksitesOrProjects;
                __with0.Columns.Clear();
                {
                    var __with1 = __with0.Columns;
                    __with1.Add("Nummer:");
                    __with1.Add("Arbeitsplatzname:");
                }
            }
        }

        private void RebuildLists()
        {
            lvwLegatroWorksitesOrProjects.Items.Clear();
            tvwAssignments.Nodes.Clear();
            if (!(string.IsNullOrEmpty(myTaskItem.LegatroSQLConnectionString)))
            {
                LegatroDataContext dc = new LegatroDataContext(myTaskItem.LegatroSQLConnectionString);
                try
                {
                    myWorksiteOrProjects = ((
                        from wsItem in dc.WorksitesOrProjects
                        where !(wsItem.IsProject)orderby wsItem.WorkEntityNumber
                        select wsItem)).ToList();
                }
                catch (Exception ex)
                {
                    MessageBox.Show("Beim Abrufen der Daten aus der Legatro-Datenbank ist ein Fehler aufgetreten." + System.Environment.NewLine + "Bitte Überprüfen Sie die Netzwerkverbindung und die Verbindungszeichenfolge zur Datenbank.", "Verbindungsaufbau zu Legatro nicht möglich!", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                foreach (var wsItem in myWorksiteOrProjects)
                {
                    ListViewItem lvwItem = new ListViewItem(wsItem.WorkEntityNumber.ToString("000000"));
                    lvwItem.SubItems.Add(wsItem.WorkEntityName);
                    lvwItem.Tag = wsItem;
                    //Rausfinden, ob das Item schon zugewiesen wurde
                    if (myTaskItem.ConversionItems != null)
                    {
                        foreach (var taskItem in myTaskItem.ConversionItems)
                        {
                            if (taskItem.AlienElementID == wsItem.WorkEntityNumber)
                            {
                                //...dann in Fettschrift anzeigen
                                if (taskItem.HomeElementID > -1)
                                {
                                    lvwItem.Font = new System.Drawing.Font(lvwItem.Font, System.Drawing.FontStyle.Bold);
                                }
                            }
                        }
                    }

                    lvwLegatroWorksitesOrProjects.Items.Add(lvwItem);
                }
            }

            foreach (ColumnHeader colItem in lvwLegatroWorksitesOrProjects.Columns)
            {
                colItem.Width = -2;
            }

            WorkGroupInfoItems wgItems = new WorkGroupInfoItems(false);
            mySortedWorkgroupInfoItems = ((
                from wgSortedItem in wgItems
                orderby wgSortedItem.WorkGroupNumber
                select wgSortedItem)).ToList();
            foreach (var item in mySortedWorkgroupInfoItems)
            {
                var tn = tvwAssignments.Nodes.Add(item.WorkGroupNumber + ": " + item.WorkGroupName);
                tn.Tag = item;
                if (item.IsActive)
                {
                    tn.ForeColor = System.Drawing.Color.Blue;
                    tn.NodeFont = new System.Drawing.Font(this.Font, System.Drawing.FontStyle.Bold);
                }
                else
                {
                    tn.ForeColor = System.Drawing.Color.Red;
                }
            }

            Application.DoEvents();
            if (myTaskItem.ConversionItems == null)
            {
                myTaskItem.ConversionItems = myTaskItem.AssembleConversionItems();
            }
            else
            {
                //Abgleich, ob sich was geändert hat
                var tmpItems = myTaskItem.AssembleConversionItems();
                var itemsToAdd = new List<FacessoConversionItemBase>();
                foreach (var tmpItem in tmpItems)
                {
                    var found = false;
                    foreach (var taskItem in myTaskItem.ConversionItems)
                    {
                        if (taskItem.AlienElementID == tmpItem.AlienElementID)
                        {
                            found = true;
                            break;
                        }
                    }

                    if (!(found))
                    {
                        itemsToAdd.Add(tmpItem);
                    }
                }

                foreach (var itemToAdd in itemsToAdd)
                {
                    myTaskItem.ConversionItems.Add(itemToAdd);
                }
            }

            foreach (var item in myTaskItem.ConversionItems)
            {
                //Ein Facesso-Element (Worksite) ist zugeordnet bei >-1
                if (item.HomeElementID > -1)
                {
                    //Home-Element im Tree suchen
                    foreach (TreeNode nodeItem in tvwAssignments.Nodes)
                    {
                        if (((WorkGroupInfo)nodeItem.Tag).WorkGroupNumber == item.HomeElementID)
                        {
                            //Gefunden: Alien-Element und Item-Name als Zwei eintragen
                            var tmpNode = nodeItem.Nodes.Add(item.AlienElementID.ToString("000000") + ": " + item.Itemname);
                            tmpNode.Tag = item.HomeElementID;
                            nodeItem.Expand();
                        }
                    }
                }
            }
        }

        private void btnAdd_Click(System.Object sender, System.EventArgs e)
        {
            if (lvwLegatroWorksitesOrProjects.SelectedItems.Count == 0)
            {
                MessageBox.Show("Bitte wählen Sie einen Legatro-Arbeitsplatz zum Zuweisen aus!", "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            if (tvwAssignments.SelectedNode == null)
            {
                MessageBox.Show("Bitte wählen Sie eine Facesso-Produktiv-Site aus, der der Legatro-Arbeitsplatz zugewiesen werden soll!", "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            var legItem = ((WorksitesOrProjects)lvwLegatroWorksitesOrProjects.SelectedItems[0].Tag);
            var facItem = ((WorkGroupInfo)tvwAssignments.SelectedNode.Tag);
            foreach (var cItem in myTaskItem.ConversionItems)
            {
                if (cItem.AlienElementID == legItem.WorkEntityNumber)
                {
                    cItem.HomeElementID = facItem.WorkGroupNumber;
                    cItem.HomeElementName = facItem.WorkGroupName;
                    break;
                }
            }

            RebuildLists();
        }

        private void btnRemove_Click(System.Object sender, System.EventArgs e)
        {
            if (tvwAssignments.SelectedNode == null)
            {
                MessageBox.Show("Bitte wählen Sie eine Facesso-Produktiv-Site aus, deren Zuweisung aufgehoben werden soll!", "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            if (tvwAssignments.SelectedNode.Parent == null & tvwAssignments.SelectedNode.Nodes.Count == 0)
            {
                MessageBox.Show("Bitte wählen Sie eine Zweig aus, der ein Legatro-Arbeitsplatz/Kostenstelle zugeordnet ist!", "Fehlende Auswahl", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            TreeNode currentNode = default(TreeNode);
            //Dafür sorgen, dass immer der Hauptzweig ausgewählt ist.
            if (tvwAssignments.SelectedNode.Parent != null)
            {
                currentNode = tvwAssignments.SelectedNode.Parent;
            }
            else
            {
                currentNode = tvwAssignments.SelectedNode;
            }

            //Zugeordnetes WorkgroupInfo-Objekt finden
            var workgroupNumber = System.Convert.ToInt32(currentNode.Nodes[0].Tag);
            var facItem = ((
                from fItem in mySortedWorkgroupInfoItems
                where fItem.WorkGroupNumber == workgroupNumber
                select fItem)).SingleOrDefault();
            if (facItem == null)
            {
                MessageBox.Show("Interner Zuordnungsfehler - bitte rufen Sie nach Verlassen diesen Dialog erneut auf.", "Interner Zuordnungsfehler", MessageBoxButtons.OK, MessageBoxIcon.Asterisk);
                return;
            }

            //Sicherheitsabfrage:
            var dr = MessageBox.Show("Sind Sie sicher, dass Sie die Zuordnung zur Produktiv-Site " + facItem.DisplayName + " aufheben wollen?", "Zuordnung aufheben?", MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button2);
            if (dr == System.Windows.Forms.DialogResult.Yes)
            {
                foreach (var cItem in myTaskItem.ConversionItems)
                {
                    if (cItem.HomeElementID == facItem.WorkGroupNumber)
                    {
                        cItem.HomeElementID = -1;
                        cItem.HomeElementName = "";
                        break;
                    }
                }

                RebuildLists();
            }
        }

        public frmLegatroTimeDataConfigDialog()
        {
            InitializeComponent();
        }
    }
}